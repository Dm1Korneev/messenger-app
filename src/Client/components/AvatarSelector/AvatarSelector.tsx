import ClearIcon from '@mui/icons-material/Clear';
import CloudUploadIcon from '@mui/icons-material/CloudUpload';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import InputLabel from '@mui/material/InputLabel';
import { useEffect, useState } from 'react';

import { UsersAvatar } from '../UsersAvatar';

type Props = {
  avatarFileInput: React.RefObject<HTMLInputElement>;
  onChange?: () => void;
  avatar?: string;
  disabled?: boolean;
}

export const AvatarSelector = ({
  avatar: avatarProp, onChange = () => {}, avatarFileInput, disabled = false,
}: Props) => {
  const [avatar, setAvatar] = useState<string>();

  useEffect(() => {
    setAvatar(avatarProp);
  }, [avatarProp]);

  const avatarChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const files = event?.target?.files;
    if (!files || !files.length) {
      return;
    }

    const newAvatar = URL.createObjectURL(files[0]);

    setAvatar(newAvatar);
    onChange();
  };

  const handlerRemoveButton = (event: React.MouseEvent) => {
    event.preventDefault();
    setAvatar(undefined);
    onChange();
  };

  return (
    <Box display="flex" mt={2}>
      <Box mr={2}>
        <UsersAvatar avatar={avatar} size={60} />
      </Box>
      <Box display="flex" flexDirection="column">
        <InputLabel disabled={disabled}>
          <span>Avatar</span>
          <input
            ref={avatarFileInput}
            accept="image/*"
            onChange={avatarChange}
            style={{ display: 'none' }}
            type="file"
          />
          <Box display="flex" flexDirection="row" mt={1}>
            <Button
              component="span"
              disabled={disabled}
              size="small"
              sx={{ mr: 1 }}
              variant="outlined"
            >
              <CloudUploadIcon sx={{ mr: 1 }} />
              Upload
            </Button>
            <Button
              component="span"
              disabled={disabled}
              onClick={handlerRemoveButton}
              size="small"
              sx={{ mr: 1 }}
              variant="outlined"
            >
              <ClearIcon sx={{ mr: 1 }} />
              Remove
            </Button>
          </Box>
        </InputLabel>
      </Box>
    </Box>
  );
};
