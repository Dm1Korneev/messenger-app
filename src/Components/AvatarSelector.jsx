import React from 'react';
import PropTypes from 'prop-types';

import withStyles from '@material-ui/core/styles/withStyles';
import InputLabel from '@material-ui/core/InputLabel';
import Button from '@material-ui/core/Button';
import CloudUploadIcon from '@material-ui/icons/CloudUpload';
import ClearIcon from '@material-ui/icons/Clear';

import UsersAvatar from 'Components/UsersAvatar';

class AvatarSelector extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      avatar: props.avatar,
    };
  }

  avatarChange = (event) => {
    if (!event.target.files.length) {
      return;
    }

    const avatar = URL.createObjectURL(event.target.files[0]);

    this.setState({
      avatar,
    });
  };

  handleOnChange = (event) => {
    const { onChange } = this.props;

    this.avatarChange(event);
    if (onChange) {
      onChange(event);
    }
  };

  handlerRemoveButton = (event) => {
    const { onChange, avatarFileInput } = this.props;

    event.preventDefault();
    avatarFileInput.current.value = '';
    this.setState({
      avatar: undefined,
    });
    if (onChange) {
      onChange(event);
    }
  };

  render() {
    const { avatar } = this.state;
    const { classes, avatarFileInput } = this.props;

    return (
      <div className={classes.avatarSelector}>
        <div className={classes.avatar}>
          <UsersAvatar avatar={avatar} size={60} />
        </div>
        <div className={classes.avatarSelector__labalButtons}>
          <InputLabel>
            <span>Avatar</span>
            <input
              accept="image/*"
              style={{ display: 'none' }}
              onChange={this.handleOnChange}
              type="file"
              ref={avatarFileInput}
            />
            <div className={classes.avatarSelector__buttons}>
              <Button
                variant="outlined"
                component="span"
                size="small"
                className={classes.button}
              >
                <CloudUploadIcon className={classes.leftIcon} />
                Upload
              </Button>
              <Button
                variant="outlined"
                component="span"
                size="small"
                className={classes.button}
                onClick={this.handlerRemoveButton}
              >
                <ClearIcon className={classes.leftIcon} />
                Remove
              </Button>
            </div>
          </InputLabel>
        </div>
      </div>
    );
  }
}

AvatarSelector.propTypes = {
  classes: PropTypes.instanceOf(Object).isRequired,
  avatarFileInput: PropTypes.instanceOf(Object).isRequired,
  onChange: PropTypes.func,
  avatar: PropTypes.string,
};

AvatarSelector.defaultProps = {
  onChange: () => {},
  avatar: undefined,
};

const styles = (theme) => ({
  avatar: { marginRight: theme.spacing.unit * 2 },
  avatarSelector: {
    marginTop: theme.spacing.unit * 2,
    display: 'flex',
  },
  avatarSelector__labalButtons: {
    display: 'flex',
    flexDirection: 'column',
  },
  leftIcon: {
    marginRight: theme.spacing.unit,
  },
  avatarSelector__buttons: {
    marginTop: theme.spacing.unit,
    display: 'flex',
    flexDirection: 'row',
  },
  button: {
    marginRight: theme.spacing.unit,
  },
});

export default withStyles(styles)(AvatarSelector);
