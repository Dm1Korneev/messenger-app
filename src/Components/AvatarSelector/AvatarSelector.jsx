import React, { useEffect, useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import InputLabel from '@material-ui/core/InputLabel';
import Button from '@material-ui/core/Button';
import CloudUploadIcon from '@material-ui/icons/CloudUpload';
import ClearIcon from '@material-ui/icons/Clear';

import UsersAvatar from 'Components/UsersAvatar';

const useStyles = makeStyles((theme) => ({
  avatar: { marginRight: theme.spacing(2) },
  avatarSelector: {
    marginTop: theme.spacing(2),
    display: 'flex',
  },
  avatarSelector__labalButtons: {
    display: 'flex',
    flexDirection: 'column',
  },
  leftIcon: {
    marginRight: theme.spacing(1),
  },
  avatarSelector__buttons: {
    marginTop: theme.spacing(1),
    display: 'flex',
    flexDirection: 'row',
  },
  button: {
    marginRight: theme.spacing(1),
  },
}));

const AvatarSelector = ({
  avatar: avatarProp, onChange, avatarFileInput, disabled,
}) => {
  const classes = useStyles();
  const [avatar, setAvatar] = useState();

  useEffect(() => {
    setAvatar(avatarProp);
  }, [avatarProp]);

  const avatarChange = (event) => {
    if (!event.target.files.length) {
      return;
    }

    const newAvatar = URL.createObjectURL(event.target.files[0]);

    setAvatar(newAvatar);
    if (onChange) {
      onChange(event);
    }
  };

  const handlerRemoveButton = (event) => {
    event.preventDefault();
    setAvatar(undefined);
    if (onChange) {
      onChange(undefined);
    }
  };

  return (
    <div className={classes.avatarSelector}>
      <div className={classes.avatar}>
        <UsersAvatar avatar={avatar} size={60} />
      </div>
      <div className={classes.avatarSelector__labalButtons}>
        <InputLabel disabled={disabled}>
          <span>Avatar</span>
          <input
            accept="image/*"
            style={{ display: 'none' }}
            onChange={avatarChange}
            type="file"
            ref={avatarFileInput}
          />
          <div className={classes.avatarSelector__buttons}>
            <Button
              variant="outlined"
              component="span"
              size="small"
              className={classes.button}
              disabled={disabled}
            >
              <CloudUploadIcon className={classes.leftIcon} />
              Upload
            </Button>
            <Button
              variant="outlined"
              component="span"
              size="small"
              className={classes.button}
              onClick={handlerRemoveButton}
              disabled={disabled}
            >
              <ClearIcon className={classes.leftIcon} />
              Remove
            </Button>
          </div>
        </InputLabel>
      </div>
    </div>
  );
};

AvatarSelector.propTypes = {
  avatarFileInput: PropTypes.instanceOf(Object).isRequired,
  onChange: PropTypes.func,
  avatar: PropTypes.string,
  disabled: PropTypes.bool,
};

AvatarSelector.defaultProps = {
  onChange: () => {},
  avatar: undefined,
  disabled: false,
};

export default AvatarSelector;
