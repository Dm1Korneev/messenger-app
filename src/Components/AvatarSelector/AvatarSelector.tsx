import React, { useEffect, useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import InputLabel from '@material-ui/core/InputLabel';
import Button from '@material-ui/core/Button';
import CloudUploadIcon from '@material-ui/icons/CloudUpload';
import ClearIcon from '@material-ui/icons/Clear';

import UsersAvatar from 'Components/UsersAvatar';

const useStyles = makeStyles((theme) => ({
  avatar: { marginRight: theme.spacing(2) },
  avatarSelector: {
    marginTop: theme.spacing(2),
    display: 'flex',
  },
  avatarSelector__labalButtons: {
    display: 'flex',
    flexDirection: 'column',
  },
  leftIcon: {
    marginRight: theme.spacing(1),
  },
  avatarSelector__buttons: {
    marginTop: theme.spacing(1),
    display: 'flex',
    flexDirection: 'row',
  },
  button: {
    marginRight: theme.spacing(1),
  },
}));

type Props = {
  avatarFileInput: React.RefObject<HTMLInputElement>;
  onChange?: () => void;
  avatar?: string;
  disabled?: boolean;
}

const AvatarSelector = ({
  avatar: avatarProp, onChange = () => {}, avatarFileInput, disabled = false,
}: Props) => {
  const classes = useStyles();
  const [avatar, setAvatar] = useState<string>();

  useEffect(() => {
    setAvatar(avatarProp);
  }, [avatarProp]);

  const avatarChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const files = event?.target?.files;
    if (!files || !files.length) {
      return;
    }

    const newAvatar = URL.createObjectURL(files[0]);

    setAvatar(newAvatar);
    onChange();
  };

  const handlerRemoveButton = (event: React.MouseEvent) => {
    event.preventDefault();
    setAvatar(undefined);
    onChange();
  };

  return (
    <div className={classes.avatarSelector}>
      <div className={classes.avatar}>
        <UsersAvatar avatar={avatar} size={60} />
      </div>
      <div className={classes.avatarSelector__labalButtons}>
        <InputLabel disabled={disabled}>
          <span>Avatar</span>
          <input
            accept="image/*"
            style={{ display: 'none' }}
            onChange={avatarChange}
            type="file"
            ref={avatarFileInput}
          />
          <div className={classes.avatarSelector__buttons}>
            <Button
              variant="outlined"
              component="span"
              size="small"
              className={classes.button}
              disabled={disabled}
            >
              <CloudUploadIcon className={classes.leftIcon} />
              Upload
            </Button>
            <Button
              variant="outlined"
              component="span"
              size="small"
              className={classes.button}
              onClick={handlerRemoveButton}
              disabled={disabled}
            >
              <ClearIcon className={classes.leftIcon} />
              Remove
            </Button>
          </div>
        </InputLabel>
      </div>
    </div>
  );
};

export default AvatarSelector;
