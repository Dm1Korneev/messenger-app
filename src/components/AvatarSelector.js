import React from "react";
import withStyles from "@material-ui/core/styles/withStyles";
import InputLabel from "@material-ui/core/InputLabel";
import UsersAvatar from "./UsersAvatar";
import Button from "@material-ui/core/Button";
import CloudUploadIcon from "@material-ui/icons/CloudUpload";

class AvatarSelector extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      avatar: undefined
    };
  }

  handlerAvatarChange = event => {
    if (!event.target.files.length) {
      return;
    }

    const avatar = URL.createObjectURL(event.target.files[0]);

    this.setState({
      avatar
    });
  };

  render() {
    const { avatar } = this.state;
    const { classes, avatarFileInput } = this.props;

    return (
      <div className={classes.avatarSelector}>
        <div className={classes.avatar}>
          <UsersAvatar avatar={avatar} size={60} />
        </div>
        <div className={classes.avatarSelector__labalButtons}>
          <InputLabel>
            <span>Avatar</span>
            <input
              accept="image/*"
              style={{ display: "none" }}
              onChange={this.handlerAvatarChange}
              type="file"
              ref={avatarFileInput}
            />
            <div className={classes.avatarSelector__buttons}>
              <Button variant="outlined" component="span" size="small">
                <CloudUploadIcon className={classes.leftIcon} />
                Upload
              </Button>
            </div>
          </InputLabel>
        </div>
      </div>
    );
  }
}

const styles = theme => ({
  avatar: { marginRight: theme.spacing.unit * 2 },
  avatarSelector: {
    marginTop: theme.spacing.unit * 2,
    display: "flex"
  },
  avatarSelector__labalButtons: {
    display: "flex",
    flexDirection: "column"
  },
  leftIcon: {
    marginRight: theme.spacing.unit
  },
  avatarSelector__buttons: {
    marginTop: theme.spacing.unit,
    display: "flex",
    flexDirection: "row"
  }
});

export default withStyles(styles)(AvatarSelector);
