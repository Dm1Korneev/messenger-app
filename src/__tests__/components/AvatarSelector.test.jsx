import React from 'react';

import AvatarSelector from 'Components/AvatarSelector';

jest.mock('Components/UsersAvatar', () => global.mockComponent('UsersAvatar'));

jest.mock('@material-ui/core/InputLabel', () => global.mockComponent('InputLabel'));
jest.mock('@material-ui/core/Button', () => global.mockComponent('Button'));
jest.mock('@material-ui/icons/CloudUpload', () => global.mockComponent('CloudUploadIcon'));
jest.mock('@material-ui/icons/Clear', () => global.mockComponent('ClearIcon'));

const props = {
  avatarFileInput: React.createRef(),
};

describe('render AvatarSelector component', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = global.mount(
      <AvatarSelector {...props} />,
    );
  });

  test('component is render', () => {
    expect(wrapper.find('AvatarSelector').length).toBe(1);
  });

  test('UsersAvatar subcomponent is render', () => {
    expect(wrapper.find({ originalcomponent: 'UsersAvatar' }).length).toBe(1);
  });

  test('InputLabel subcomponent is render', () => {
    expect(wrapper.find({ originalcomponent: 'InputLabel' }).length).toBe(1);
  });
});

describe('snapshot-test AvatarSelector component', () => {
  test('Renders correct properties', () => {
    global.mountExpect(
      <AvatarSelector {...props} />,
    ).toMatchSnapshot();
  });
});
