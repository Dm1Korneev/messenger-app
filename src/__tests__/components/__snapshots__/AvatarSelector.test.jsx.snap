// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot-test AvatarSelector component Renders correct properties 1`] = `
<WithStyles(AvatarSelector)
  avatarFileInput={
    Object {
      "current": <input
        accept="image/*"
        style="display: none;"
        type="file"
      />,
    }
  }
>
  <AvatarSelector
    avatarFileInput={
      Object {
        "current": <input
          accept="image/*"
          style="display: none;"
          type="file"
        />,
      }
    }
    classes={
      Object {
        "avatar": "AvatarSelector-avatar-1",
        "avatarSelector": "AvatarSelector-avatarSelector-2",
        "avatarSelector__buttons": "AvatarSelector-avatarSelector__buttons-5",
        "avatarSelector__labalButtons": "AvatarSelector-avatarSelector__labalButtons-3",
        "button": "AvatarSelector-button-6",
        "leftIcon": "AvatarSelector-leftIcon-4",
      }
    }
    onChange={[Function]}
  >
    <div
      className="AvatarSelector-avatarSelector-2"
    >
      <div
        className="AvatarSelector-avatar-1"
      >
        <Component
          size={60}
        >
          <div
            originalcomponent="UsersAvatar"
            testprops={
              Object {
                "avatar": undefined,
                "size": 60,
              }
            }
          />
        </Component>
      </div>
      <div
        className="AvatarSelector-avatarSelector__labalButtons-3"
      >
        <Component>
          <div
            originalcomponent="InputLabel"
            testprops={
              Object {
                "children": Array [
                  <span>
                    Avatar
                  </span>,
                  <input
                    accept="image/*"
                    onChange={[Function]}
                    style={
                      Object {
                        "display": "none",
                      }
                    }
                    type="file"
                  />,
                  <div
                    className="AvatarSelector-avatarSelector__buttons-5"
                  >
                    <Unknown
                      className="AvatarSelector-button-6"
                      component="span"
                      size="small"
                      variant="outlined"
                    >
                      <Unknown
                        className="AvatarSelector-leftIcon-4"
                      />
                      Upload
                    </Unknown>
                    <Unknown
                      className="AvatarSelector-button-6"
                      component="span"
                      onClick={[Function]}
                      size="small"
                      variant="outlined"
                    >
                      <Unknown
                        className="AvatarSelector-leftIcon-4"
                      />
                      Remove
                    </Unknown>
                  </div>,
                ],
              }
            }
          >
            <span>
              Avatar
            </span>
            <input
              accept="image/*"
              onChange={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              type="file"
            />
            <div
              className="AvatarSelector-avatarSelector__buttons-5"
            >
              <Component
                className="AvatarSelector-button-6"
                component="span"
                size="small"
                variant="outlined"
              >
                <div
                  originalcomponent="Button"
                  testprops={
                    Object {
                      "children": Array [
                        <Unknown
                          className="AvatarSelector-leftIcon-4"
                        />,
                        "Upload",
                      ],
                      "className": "AvatarSelector-button-6",
                      "component": "span",
                      "size": "small",
                      "variant": "outlined",
                    }
                  }
                >
                  <Component
                    className="AvatarSelector-leftIcon-4"
                  >
                    <div
                      originalcomponent="CloudUploadIcon"
                      testprops={
                        Object {
                          "className": "AvatarSelector-leftIcon-4",
                        }
                      }
                    />
                  </Component>
                  Upload
                </div>
              </Component>
              <Component
                className="AvatarSelector-button-6"
                component="span"
                onClick={[Function]}
                size="small"
                variant="outlined"
              >
                <div
                  originalcomponent="Button"
                  testprops={
                    Object {
                      "children": Array [
                        <Unknown
                          className="AvatarSelector-leftIcon-4"
                        />,
                        "Remove",
                      ],
                      "className": "AvatarSelector-button-6",
                      "component": "span",
                      "onClick": [Function],
                      "size": "small",
                      "variant": "outlined",
                    }
                  }
                >
                  <Component
                    className="AvatarSelector-leftIcon-4"
                  >
                    <div
                      originalcomponent="ClearIcon"
                      testprops={
                        Object {
                          "className": "AvatarSelector-leftIcon-4",
                        }
                      }
                    />
                  </Component>
                  Remove
                </div>
              </Component>
            </div>
          </div>
        </Component>
      </div>
    </div>
  </AvatarSelector>
</WithStyles(AvatarSelector)>
`;
